File: openmp/src/game-of-life.c
#include "game-of-life.h"

#include <stdlib.h>

#define IDX(i, j) ((i) * (gol->size) + (j))

void init_gol(gol_t* gol, const uint32_t grid_size, const uint32_t seed, const float density) {
    srand(seed);

    // size of the grid + 2 for the ghost cells
    gol->size = grid_size + 2;
    size_t size = (gol->size) * (gol->size) * sizeof(cell_t);

    gol->current = (cell_t*) malloc(size);
    gol->next = (cell_t*) malloc(size);

    init_grid(gol, density);
}

void init_grid(const gol_t* gol, const float density) {
    for (uint32_t i = 1; i < gol->size - 1; i++) {
        for (uint32_t j = 1; j < gol->size - 1; j++) {
            gol->current[IDX(i, j)] = (((float) rand()) / RAND_MAX < density);
        }
    }
}

uint8_t count_alive_neighbors(const gol_t* gol, const uint32_t i, const uint32_t j) {
    return  gol->current[IDX(i - 1, j - 1)] + gol->current[IDX(i - 1, j)] + gol->current[IDX(i - 1, j + 1)] +
            gol->current[IDX(i, j - 1)    ]                               + gol->current[IDX(i, j + 1)    ] +
            gol->current[IDX(i + 1, j - 1)] + gol->current[IDX(i + 1, j)] + gol->current[IDX(i + 1, j + 1)];
}

void swap_grids(gol_t* gol) {
    cell_t* temp = gol->current;
    gol->current = gol->next;
    gol->next = temp;
}

void step(const gol_t* gol) {
    for (uint32_t i = 1; i < gol->size - 1; i++) {
        for (uint32_t j = 1; j < gol->size - 1; j++) {

            uint8_t alive_neighbors = count_alive_neighbors(gol, i, j);

            // The state of the current cell
            cell_t is_alive = gol->current[IDX(i, j)];

            // The state of the current cell in the next step based on the rules of the game of life
            cell_t next_state = (is_alive && !(alive_neighbors < 2 || alive_neighbors > 3)) || (!is_alive && alive_neighbors == 3);

            gol->next[IDX(i, j)] = next_state;
        }
    }
}

void fill_ghost_cells(const gol_t* gol) {
    // Left and right borders
    for (uint32_t i = 1; i < gol->size - 1; i++) {
        gol->current[IDX(i, 0)] = gol->current[IDX(i, gol->size - 2)];
        gol->current[IDX(i, gol->size - 1)] = gol->current[IDX(i, 1)];
    }

    // Top and bottom borders
    for (uint32_t j = 0; j < gol->size; j++) {
        gol->current[IDX(0, j)] = gol->current[IDX(gol->size - 2, j)];
        gol->current[IDX(gol->size - 1, j)] = gol->current[IDX(1, j)];
    }
    
    // Corners
    gol->current[IDX(0, 0)] = gol->current[IDX(gol->size - 2, gol->size - 2)];
    gol->current[IDX(0, gol->size - 1)] = gol->current[IDX(gol->size - 2, 1)];
    gol->current[IDX(gol->size - 1, 0)] = gol->current[IDX(1, gol->size - 2)];
    gol->current[IDX(gol->size - 1, gol->size - 1)] = gol->current[IDX(1, 1)];
}

void free_gol(gol_t* gol) {
    free(gol->current);
    free(gol->next);
    free(gol);
}

File: rust/src/game_of_life.rs
use rand::Rng;
use rayon::prelude::*;

pub struct Gol {
    size: usize,
    pub current: Vec<bool>,
    pub next: Vec<bool>,
}

#[derive(Debug, Copy, Clone)]
pub struct Point {
    pub x: usize,
    pub y: usize,
}

impl Gol {
    pub fn new(grid_size: usize, density: f32) -> Gol {
        let size = grid_size + 2;
        let mut current = vec![false; (size * size) as usize];
        let next = vec![false; (size * size) as usize];

        Gol::init_grid(&mut current, size, density);

        Gol {
            size,
            current,
            next,
        }
    }

    fn init_grid(current: &mut [bool], size: usize, density: f32) {
        let mut rng = rand::thread_rng();
        for i in 1..size - 1 {
            for j in 1..size - 1 {
                current[Gol::idx(size, i, j)] = rng.gen::<f32>() < density;
            }
        }
    }

    pub fn idx(size: usize, i: usize, j: usize) -> usize {
        i * size + j
    }

    pub fn point(size: usize, index: usize) -> Point {
        Point {x: index % size, y: index / size}
    }

    fn count_alive_neighbors(&self, i: usize, j: usize) -> u8 {
        let size = self.size;
        self.current[Gol::idx(size, i - 1, j - 1)] as u8
            + self.current[Gol::idx(size, i - 1, j)] as u8
            + self.current[Gol::idx(size, i - 1, j + 1)] as u8
            + self.current[Gol::idx(size, i, j - 1)] as u8
            + self.current[Gol::idx(size, i, j + 1)] as u8
            + self.current[Gol::idx(size, i + 1, j - 1)] as u8
            + self.current[Gol::idx(size, i + 1, j)] as u8
            + self.current[Gol::idx(size, i + 1, j + 1)] as u8
    }

    pub fn swap_grids(&mut self) {
        std::mem::swap(&mut self.current, &mut self.next);
    }

    pub fn step(&mut self) {
        let size = self.size;
        for i in 1..size - 1 {
            for j in 1..size - 1 {
                let alive_neighbors = self.count_alive_neighbors(i, j);

                let is_alive = self.current[Gol::idx(size, i, j)];

                let next_state = (is_alive && !(alive_neighbors < 2 || alive_neighbors > 3))
                    || (!is_alive && alive_neighbors == 3);

                self.next[Gol::idx(size, i, j)] = next_state;
            }
        }

        self.swap_grids();
    }

    pub fn step_par(&mut self) {
        let size = self.size;

        self.next = (0..self.next.len())
            .into_par_iter()
            .map(|idx| {
                let point = Gol::point(size, idx);

                if point.x == 0 || point.x == size - 1 || point.y == 0 || point.y == size - 1 {
                    return false;
                }

                let alive_neighbors = self.count_alive_neighbors(point.x, point.y);

                let is_alive = self.current[Gol::idx(size, point.x, point.y)];

                (is_alive && !(alive_neighbors < 2 || alive_neighbors > 3)) || (!is_alive && alive_neighbors == 3)
            })
            .collect();

        self.swap_grids();
    }

    pub fn fill_ghost_cells(&mut self) {
        let size = self.size;

        // Left and right borders
        for i in 1..size - 1 {
            self.current[Gol::idx(size, i, 0)] = self.current[Gol::idx(size, i, size - 2)];
            self.current[Gol::idx(size, i, size - 1)] = self.current[Gol::idx(size, i, 1)];
        }

        // Top and bottom borders
        for j in 0..size {
            self.current[Gol::idx(size, 0, j)] = self.current[Gol::idx(size, size - 2, j)];
            self.current[Gol::idx(size, size - 1, j)] = self.current[Gol::idx(size, 1, j)];
        }

        // Corners
        self.current[Gol::idx(size, 0, 0)] = self.current[Gol::idx(size, size - 2, size - 2)];
        self.current[Gol::idx(size, 0, size - 1)] = self.current[Gol::idx(size, size - 2, 1)];
        self.current[Gol::idx(size, size - 1, 0)] = self.current[Gol::idx(size, 1, size - 2)];
        self.current[Gol::idx(size, size - 1, size - 1)] = self.current[Gol::idx(size, 1, 1)];
    }
}